opt: index accelerate && (overlaps) expressions for array inverted indexes

RajivTS opened the pull request:
Previously, we did not support index acceleration for queries involving
array overlaps (&&).

This change adds support for using the inverted index with && expressions on
Array columns. When there is an inverted index available, a scan will be done on
the Array column using the spans found from the constant value.

Fixes: #75477

Release note (performance improvement): Expressions using the overlaps (&&)
operator for arrays now support index-acceleration for faster execution in
some cases.

Release justification: low risk, high benefit changes to existing
functionality.

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=77418) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=77418) before we can accept your contribution.<br/><hr/>**Rajiv Sharma** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=77418) it.</sub>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/77418)
<!-- Reviewable:end -->

rytaft reviewed (approved):
Sorry for the delay, I was OOO. Thanks for your patience!

Looks like you need to fix one of the test files to fix the CI failures -- see my comment below. Otherwise, <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 8 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77418#-:-MyAK_qV6aVb41DEQaRO:bic9h8h)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @RajivTS)

___
*[pkg/sql/rowenc/index_encoding_test.go, line 804 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77418#-MyAGyx_9n7eINMPerX4:-MyAGyx_9n7eINMPerX5:btk92eq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d74a79aff190d128ed495e4e9add062941bef424/pkg/sql/rowenc/index_encoding_test.go#L804)):*
> ```Go
> 
> 	runTest := func(indexedValue, value tree.Datum, expected, ok, unique bool) {
> 		keys, err := EncodeInvertedIndexTableKeys(indexedValue, nil, descpb.LatestNonPrimaryIndexDescriptorVersion)
> ```

Looks like this also needs to be changed to `descpb.PrimaryIndexWithStoredColumnsVersion`


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
Hi @rajivts -- we have a lot of interest from folks in this getting this functionality merged -- do you think you could update the PR so that it compiles and we can go ahead and merge it? Thank you!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77418#-:-N-AIqgR7FUeyOFca7zY:bn7swd4)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @RajivTS)



<!-- Sent from Reviewable.io -->

RajivTS committed:
opt: index accelerate && (overlaps) expressions for array inverted indexes

Previously, we did not support index acceleration for queries involving
array overlaps (&&).

This change adds support for using the inverted index with && expressions on
Array columns. When there is an inverted index available, a scan will be done on
the Array column using the spans found from the constant value.

Fixes: #75477

Release note (performance improvement): Expressions using the overlaps (&&)
operator for arrays now support index-acceleration for faster execution in
some cases.

Release justification: low risk, high benefit changes to existing
functionality.

RajivTS commented:
Apologies @rytaft, I was quite caught up with my full-time work so didn't get a chance to wrap this up. I believe its ready to go now.

rytaft reviewed (approved):
No worries! Thank you -- looks great.

bors r+

> Reviewed 6 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77418#-:-N-DmLzV7yXNjc0O2_le:b-b5jb1j)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

rytaft commented:
@rajivts in case you're curious -- here's a post from one of my colleagues who is already using this functionality! https://github.com/mgoddard/hot-fuzz 

Thanks again for your contribution!





<!-- Sent from Reviewable.io -->

RajivTS commented:
This just made my day! Thanks for sharing :)
